apps/spruce/src/pages/waterfall/useFilters.ts

```diff
@@ -0,0 +1,98 @@
+import { useMemo } from "react";
+import { WaterfallQuery, WaterfallVersionFragment } from "gql/generated/types";
+import { useQueryParam } from "hooks/useQueryParam";
+import { WaterfallFilterOptions } from "types/waterfall";
+
+export const useFilters = (waterfall: WaterfallQuery["waterfall"]) => {
+  const [requesters] = useQueryParam(
+    WaterfallFilterOptions.Requesters,
+    [] as string[],
+  );
+
+  const hasFilters = useMemo(() => requesters.length, [requesters]);
+
+  const versions = useMemo(() => {
+    if (!hasFilters) {
+      return waterfall.versions;
+    }
+
+    const filteredVersions: typeof waterfall.versions = [];
+
+    const pushInactive = (v: WaterfallVersionFragment) => {
+      if (!filteredVersions?.[filteredVersions.length - 1]?.inactiveVersions) {
+        filteredVersions.push({ version: null, inactiveVersions: [] });
+      }
+      filteredVersions[filteredVersions.length - 1].inactiveVersions?.push(v);
+    };
+
+    const pushActive = (v: WaterfallVersionFragment) => {
+      filteredVersions.push({
+        inactiveVersions: null,
+        version: v,
+      });
+    };
+
+    waterfall.versions.forEach(({ inactiveVersions, version }) => {
+      if (version) {
+        if (matchesRequesters(version, requesters)) {
+          pushActive(version);
+        } else {
+          pushInactive(version);
+        }
+      } else if (inactiveVersions) {
+        inactiveVersions.forEach(pushInactive);
+      }
+    });
+
+    return filteredVersions;
+  }, [requesters, waterfall, hasFilters]);
+
+  const activeVersionIds = useMemo(
+    () =>
+      new Set(
+        versions.reduce((ids, { version }) => {
+          if (version) {
+            ids.push(version.id);
+          }
+          return ids;
+        }, [] as string[]),
+      ),
+    [versions],
+  );
+
+  const buildVariants = useMemo(() => {
+    if (!hasFilters) {
+      return waterfall.buildVariants;
+    }
+
+    const bvs: typeof waterfall.buildVariants = [];
+    waterfall.buildVariants.forEach((bv) => {
+      if (activeVersionIds.size !== bv.builds.length) {
+        const activeBuilds: typeof bv.builds = [];
```

Comment:
nit. (same question)
```suggestion
        const activeBuilds: WaterfallBuild[] = [];
```